### PR TITLE
(small) tweaks to be JRuby friendly

### PR DESCRIPTION
**Pull request by kares**
the initial motivation is to have a real-world Rails app for testing out JRuby and its gems (such as its native AR-JDBC adapter) for fine grained compatibility and even measuring/doing performance.

---

there really was only one test failure, which was a reliance on MRI's `Time` marshaling internals, where it raises if a date is not `>= 1900` ... while this is an "incompatibility" not sure it needs to be reproduced at JRuby's end since, as no one complained about this, some might consider it a bug (and JRuby is fine marshaling a `Time` with year < 1900).

... its nicer to validate explicitly (maybe it would make sense to do more than `>= 1900` but that would be for a separate topic), details at: https://github.com/rubygems/rubygems.org/commit/3b44154cd4ef209eb27d1503e57f7c4ad747fafd (https://github.com/rubygems/rubygems.org/issues/1690)

---

there's some left-overs that do not matter much atm (to be handled elsewhere) : 

- bootsnap [does not work properly on JRuby](https://github.com/jruby/jruby/issues/5311) -> fails to boot up
- `java_count_down_latch.rb:12: warning: constant ::Fixnum is deprecated` warning from concurrent-ruby (already addressed in a 1.1 pre release)
- `rounding_helper.rb:37: warning: BigDecimal.new is deprecated` under 9.1 (fixed in 9.2)

---

last but not least this is ready for some performance testing (at least on localhost).

### CODE REVIEW

File: app/models/pusher.rb [hunk 1]
```diff
@@ -23,7 +23,9 @@ def authorize
   end
 
   def validate
-    (rubygem.valid? && version.valid?) || notify("There was a problem saving your gem: #{rubygem.all_errors(version)}", 403)
```
Comment (simi) [inline app/models/pusher.rb]: Is this change needed for JRuby to work?

**Comment by sonalkr132** [comment]
I am sorry but the utility of this change at cost of **18 more dependencies** to update doesn't seem justifiable to me. Why should rubygems.org support `jruby` anyway? The recommended way to host private servers is gemstash and it supports jruby.

**Comment by simi** [comment]
The idea behind this is to make rubygems.org usable for various JRuby testing as a real world app as @kares (JRuby and various JRuby ecosystem gems contributor/maintainer) mentioned in OP.

**Comment by sonalkr132** [comment]
I seriously doubt we are in the position to take on more dependencies to update. Maybe we can reconsider when fewer of our existing dependencies are outdated (right 60+ are outdated).

**Comment by kares** [comment]
> I am sorry but the utility of this change at cost of 18 more dependencies to update doesn't seem justifiable to me. Why should rubygems.org support jruby anyway? The recommended way to host private servers is gemstash and it supports jruby.

Thanks Aditya, if you look at *Gemfile.lock* there's pretty much 4 "more" dependencies : 
- **activerecord-jdbc-adapter** (= 51.2) the AR adapter for JRuby
- **jdbc-postgres** (>= 9.4, < 43) the PG driver
- **puma** used as a multi-threaded server opposed to a forking one (for testing and development)
- and **spoon** which comes in as a dependecy of launchy on JRuby

the rest are -java versions of gems that exist in and in past versions Bundler wouldn't even care to list them in the .lock file (but its better that Bundler is now explicit) - they will be updated along side under MRI without much hustle. I am not sure that would cause any headaches for the RG.org team? you do not need to care for AR-JDBC adapter stuff, since you do not run on production - the only (minor) annoyance there will be updating when going for a new major Rails version.
maybe the CI could 'slow' builds down or going red due an incompatibility, but we can set JRuby for the `allowed_failures:` section and there aren't that many builds during a day.

as to why support JRuby, it seemed like a really simple Rails app that works out-of-the-box fine with JRuby.
so the JRuby team would welcome being able to do load testing against the up-to-date version of the app. there might be gains for the future, can imagine it could cut the production bill considerably, although that is irrelevant atm :wink:

**Comment by kares** [comment]
will do a rebase, as all of the initial left-overs should now be irrelevant with JRuby 9.2 : 

>    bootsnap does not work properly on JRuby -> fails to boot up
>    java_count_down_latch.rb:12: warning: constant ::Fixnum is deprecated warning from concurrent-ruby (already addressed in a 1.1 pre release)
>    rounding_helper.rb:37: warning: BigDecimal.new is deprecated under 9.1 (fixed in 9.2)

certainly do not want to impose any burden on the RGs team, happy you guys were able to take a look

**Comment by indirect** [comment]
Can you say more about why you opened this PR? Are you running a copy of RubyGems.org in production on JRuby somewhere? The RubyGems.org team doesn't use JRuby, and doesn't officially support running this Rails app anywhere except the RubyGems.org servers.

**Comment by simi** [comment]
@indirect I think the answer lives in first part of first post.

> the initial motivation is to have a real-world Rails app for testing out JRuby and its gems (such as its native AR-JDBC adapter) for fine grained compatibility and even measuring/doing performance.

**Comment by indirect** [comment]
@simi good point, thanks! After some initial conversations with the rest of the rubygems.org team, it sounds like we don't have the resources to officially support JRuby or add it to the CI build, but accepting PRs that fix JRuby issues without breaking MRI seems likely fine. 👍

**Comment by simi** [comment]
Currently it is in `allow_failures` and it can be removed at any time if it will be unmaintained.

**Comment by kares** [comment]
also fine to just comment out CI on JRuby (as its not officially supported), 
let me adjust the PR in hopes of getting the bits to be able to run the app against JRuby wout a hustle.

**Comment by colby-swandale** [comment]
Hi @kares!

We've discussed this issue internally and agreed as @indirect has pointed out - that we're happy to accept small tweaks to support interpreters outside of MRI. But, we won't be accepting new gems specifically to support JRuby or adding JRuby into our CI matrix at this time.

**Comment by kares** [comment]
Thanks Colby, wout the added gem we won't be able to really play with this app using JRuby.
JRuby on Rails apps simply use replacement gems for ActiveRecord's native (C) DB adapters.
